src/function.rs

```diff
@@ -661,6 +662,61 @@ fn uuid(_context: Context) -> FunctionResult {
   Ok(uuid::Uuid::new_v4().to_string())
 }
 
+fn which(context: Context, s: &str) -> FunctionResult {
```

Comment:
What do you think about this version? I found the original logic a little hard to follow:

(Also included some comments which don't need to go into the final PR.)

```rust
fn which(context: Context, command: &str) -> FunctionResult {
  use std::path::Component;

  let command = Path::new(command);

  let relative = match command.components().next() {
    None => return Err("empty command".into()),
    // Any path that starts with `.` or `..` can't be joined to elements of `$PATH` and should be considered on its own. (Is this actually true? What about `C:foo` on windows? Is that a thing?
    Some(Component::CurDir) | Some(Component::ParentDir) => vec![command.into()],
    _ => {
      let paths = env::var_os("PATH").ok_or("`PATH` environment variable not set")?;

      env::split_paths(&paths)
        .map(|path| path.join(command))
        .collect()
    }
  };

  let working_directory = context.evaluator.context.working_directory();

  let absolute = relative
    .into_iter()
    // note that an path.join(absolute_path) winds up being absolute_path
    // lexiclean is hear to remove unnecessary `.` and `..`
    .map(|relative| working_directory.join(relative).lexiclean())
    .collect::<Vec<PathBuf>>();

  for candidate in absolute {
    if is_executable::is_executable(&candidate) {
      return candidate
        .to_str()
        .map(str::to_string)
        .ok_or_else(|| format!("Executable path not unicode: {}", candidate.display()));
    }
  }

  Ok(String::new())
}
```

Reply:
I personally find my implementation easier to understand, but that might only be because I wrote it.

The main difference between our implementations seems to be how we determine whether to resolve via `PATH`. Mine checks `command.components.count()`: anything with more than one component is considered a relative path and skips `PATH` resolution. Meanwhile, yours checks `command.components.next()`: anything whose first element is `.`, `..`, or `/` is considered a relative path and skips `PATH` resolution.

Crucially, they differ on how they handle something like `which("dir/cmd")`, which has more than one component, but the first component is a `Component::Normal("dir")`. Thus, my implementation would consider it the same as `./dir/cmd` (i.e., resolved relative to CWD), while yours would resolve it using the `PATH` variable.

I wrote a small test, and (at least on macOS with `sh`, `bash`, and `zsh`) it seems like my implementation has the correct behavior in this instance:

```sh
#!/bin/sh

set -e
tmpdir="$(mktemp -d)"
cd "$tmpdir"

mkdir -p path/dir dir

printf "#!%s\necho resolved via PATH" "$SHELL" > path/dir/cmd
chmod +x path/dir/cmd

printf "#!%s\necho resolved via CWD" "$SHELL" > dir/cmd
chmod +x dir/cmd

PATH="$(realpath path)" dir/cmd   # prints 'resolved via CWD'

rm -rf "$tmpdir"
```

The other differences are:
- You use `Path::new(command)` rather than `PathBuf::from(command)`. I've adopted this change since it's more efficient (and also 
- You use `lexiclean()` to remove extraneous `..` or `.` components. However, at least the `which` implementation on my system (macOS) does not do this.
- You handle absolute paths (those beginning with `/`) implicitly, through the behavior of `Path::join()` (TIL about this behavior!). But I personally think it's clearer to handle this case explicitly.
- You make two passes through the candidates, one to prepend the working directory, and one to check whether it is an executable. I didn't do this to avoid making another copy.